### ISSUE TITLE
Debugging

### ISSUE THREAD

**Issue by PythonLinks** [iodide-project/pyodide#550]
A killer app for PyOdide would be the educational market.  Where students would not have to install python on their machine, just use PyOdide in the browser. 

The problem is that they need a debugger.  Not a web assembly debugger, but a Python debugger. is this possible?  What would I need to do to make it work?  Does it already work?  Would IPython solve the problems?

I am pretty motivated to make this work, but I just do not know what I would need to do. 

For context, I  am already editing Python in the Browser, and using the PyOdide compile statement  to do the syntax checking.  
https://demo.forestwiki.com/documentation/HtmlContentTypes/python.py/acedemo

One user pointed out that I just need to add a Run button to make it great for education.  But I also need a debug option.  

So how can I debug Python running in PyOdide?

**Comment by rth** [comment]
It looks like currently [pdb tests fail](https://github.com/iodide-project/pyodide/blob/55857ffcac9e20f193032682d0d16c73b2975c67/test/python_tests.txt#L352) because of the subprocess module. It could be worth investigating what are the exact failure there, and if there are any workarounds. Alternatively maybe there is a way to run a python debugger with threading (once https://github.com/iodide-project/pyodide/issues/237 is resolved) -- not sure.

**Comment by PythonLinks** [comment]
I do not think that I am smart enough to fix cPython on WebAssembly.  Too much I do not know.  
One guy at EuroPython built a patch for cPython, but could not get the tests to run.   Reportedly PyPy.js runs a repl in the browser, so I will try that.  They are pretty good at being cPython compatible.  PyPy.js does not have the PyOdide libraries, but for the educational market that is fine.  In due course someone much smarter than me will fix this problem.

**Comment by PythonLinks** [comment]
Here is what happens when you run pdb in PyOdide.

This discussion is copied from 
https://github.com/ethanchewy/PythonBuddy/issues/46

Demo'd on @dgmouris website:
https://www.stewdy.com

@dgmouris wrote: 

As for a debugger this is really wierd and I want to override the behavour of using "prompt" as I go along, debugging does work even though it's really wierd.

Example below:

def something():
    print("hi")

import pdb;pdb.set_trace()
Then click "run"

Once you have that done you can write
something()
in the prompt and then press okay.

Another prompt displays, and you can write
continue
in the prompt and press okay.

Then you press Cancel and it'll display the following pdb with the output

--Return--
> (4)()->None
hi

__________
It looks like pdb tries to grab the input from a pop up box.  If it just grabbed the input from a console, life would be great.  

I thought of using PyPy on WASM, but the github repository says the project is asleep  I can also get a reasonable REPL from skuplt.org.  So my servers will support either PyOdide with data science  libraries, but no debugger, or Skuplpt with debugger, but no data science libraries.

**Comment by laike9m** [comment]
I would be interested to know if there's a way to let users inspect Python object like they can do with Js object. For example, right now if you `print` a dict in pyodide, the output is a string:

![image](https://user-images.githubusercontent.com/2592205/68094186-2e521380-fed9-11e9-9606-864e20b3149b.png)

But if you `console.log` a JavaScript object, it outputs an object that the brower understands.

![image](https://user-images.githubusercontent.com/2592205/68094245-56da0d80-fed9-11e9-8844-3b78724f86a9.png)

For debugging, I think this kind of utility is necessary, almost all IDEs have it. With pyodide createa a full Python environment, I wouldn't expect this to be too hard.

**Comment by mdboom** [comment]
You can `console.log` a Python object and get similar behavior.

From Python:

```
from js import console
console.log({'1': 3})
```

**Comment by laike9m** [comment]
@mdboom Thanks. It works to some extent(like dict, list), but when you print an object, the output is not so idea(but that's understandable)
![image](https://user-images.githubusercontent.com/2592205/68966100-3bbea600-07d5-11ea-925e-ebdcdc6745ab.png)

![image](https://user-images.githubusercontent.com/2592205/68966082-2cd7f380-07d5-11ea-9452-0835ad119f8e.png)

I can probably convert everything to a dict on the Python side, then console.log.

**Comment by doktorschiwago** [comment]
> Here is what happens when you run pdb in PyOdide.
> 
> This discussion is copied from
> [ethanchewy/PythonBuddy#46](https://github.com/ethanchewy/PythonBuddy/issues/46)
> 
> Demo'd on @dgmouris website:
> https://www.stewdy.com
> 
> @dgmouris wrote:
> 
> As for a debugger this is really wierd and I want to override the behavour of using "prompt" as I go along, debugging does work even though it's really wierd.
> 
> Example below:
> 
> def something():
> print("hi")
> 
> import pdb;pdb.set_trace()
> Then click "run"
> 
> Once you have that done you can write
> something()
> in the prompt and then press okay.
> 
> Another prompt displays, and you can write
> continue
> in the prompt and press okay.
> 
> Then you press Cancel and it'll display the following pdb with the output
> 
> --Return--
> 
> > (4)()->None
> > hi
> 
> It looks like pdb tries to grab the input from a pop up box. If it just grabbed the input from a console, life would be great.
> 
> I thought of using PyPy on WASM, but the github repository says the project is asleep I can also get a reasonable REPL from skuplt.org. So my servers will support either PyOdide with data science libraries, but no debugger, or Skuplpt with debugger, but no data science libraries.

I think the behaviour is due to emscripten stdin default behaviour. It is not limited to pdb as you can see by entering 

name = input("Enter your name: ") 

into the repl. See chapter "Setting up standard I/O devices" in https://emscripten.org/docs/api_reference/Filesystem-API.html for stdin defaulting to window.prompt.

**Comment by johnwason** [comment]
Have you tried passing a custom file object implementation to `stdin` and `stdout` in the constructor of `Pdb`?

https://docs.python.org/3/library/pdb.html#pdb.Pdb

It should be straightforward to connect a custom file object to JS messages or maybe a callback of some sort, which can proxy the input and output from some type of console. The console can be something custom, or if possible integrate directly into the browser debug console. I don't know how flexible the browser debuggers are.

**Comment by johnwason** [comment]
Another possible solution is to use a "reverse attach" remote debugger. Reverse attach would open a websocket and connect to the debugger. This is "reverse" because typically the debugger would be the client and the program with debugging would be the server. In this case, the program running in the browser will initiate the connection. debugpy running in normal mode seems to support reverse connections. Maybe it can be modified to run in the browser and use websockets.

https://github.com/microsoft/debugpy

**Comment by hoodmane** [comment]
Well with `Atomics` on a webworker you can make stdin work correctly. It'd be best to do this for the entire application. For instance I think pdb works fine in my demo version here:
https://spectralsequences.com/worker-pyodide-console/

Unfortunately this doesn't work in Safari:
https://caniuse.com/?search=Atomics

> I don't know how flexible the browser debuggers are.

Not flexible enough for this unfortunately.

**Comment by hoodmane** [comment]
Adjusting the console representation of the logging output is possible in chrome with custom formatters:
https://www.mattzeunert.com/2016/02/19/custom-chrome-devtools-object-formatters.html

**Comment by johnwason** [comment]
The websocket method would probably be the websocket method. It might require a small Python server to proxy the data from the websocket format to a format that an IDE would understand. Eclipse and VS Code are using PyDevd so I don't know if they would be compatible with pdb.

**Comment by johnwason** [comment]
I meant to say the websocket method would probably be the fastest way to get debugging up and running.

**Comment by alexmojaki** [comment]
Just wanted to share my project http://futurecoder.io/ which is working smoothly with Pyodide:

- In a webworker
- Using Comlink
- With input() and other stdin working, including pdb
- In a shell styled UI, no alert() dialogs
- Using `SharedArrayBuffer` and `Atomics.wait` for stdin (sorry Safari users)
- Also using `SharedArrayBuffer` for interrupting, but it isn't working in 0.18.0 so I need #1797 to merge before I can upgrade
- Used @hoodmane's worker pyodide console repo to figure it out, thanks for that! BTW https://spectralsequences.com/worker-pyodide-console/ isn't working any more because of the interrupt issue.
- Plus futurecoder also has other integrated Python debuggers [snoop](https://github.com/alexmojaki/snoop) and [birdseye](https://github.com/alexmojaki/birdseye) which run in the same worker with Pyodide, along with enhanced tracebacks and other features that all rely heavily on CPython internals, so this would not work at all with other browser interpreters such as Skulpt. For that, a **HUGE THANK YOU** to the whole Pyodide team for making this possible!!

Pinging users from related issues who I think would be interested: @mfripp @stefnotch @joemarshall

**Comment by hoodmane** [comment]
@alexmojaki If `setInterruptBuffer` is your only blocker for the upgrade to 0.18, you could use `pyodide._module.setInterruptBuffer` which does work.

**Comment by joemarshall** [comment]
@alexmojaki That is super cool - I wish my users didn't have Macs or iPhones, so that the sharedarraybuffer stuff worked. Then I wouldn't have to mess around with unthrowing exceptions to be able to read console input etc.

**Comment by rth** [comment]
> Just wanted to share my project http://futurecoder.io 

@alexmojaki Looks great! Would you be interested in adding it to https://github.com/pyodide/pyodide/blob/main/docs/project/related-projects.md ?

**Comment by alexmojaki** [comment]
> @alexmojaki is by a miracle your project open source?

Yes, the repo is linked from the site, do you think I should make it more visible?

> I tried what @johnwason proposed but I am still getting the alert window.

I think that implies you're running Pyodide in the main thread, the first step in all this is to run it in a web worker.

> All I want is to redirect the stdin to the pyodide console.

Unfortunately it's not as easy as it sounds, although I've tried to make it easier. Check out https://github.com/alexmojaki/pyodide-worker-runner
